### ISSUE TITLE
Move custom buttons automatically to the sidebar

### ISSUE THREAD

**Issue by salex3** [stephanrauh/ngx-extended-pdf-viewer#1584]
We have several custom buttons implemented. 

We've noticed that the provided button components, usually hide in the sidebar.

That was no issue for us, until we decided we're going to make our website into a hybrid app, and we have enabled several features, plus our custom buttons.

<img width="929" alt="Screenshot 2022-11-17 at 10 03 24" src="https://user-images.githubusercontent.com/14350155/202390329-d73e5410-9bff-4db6-8713-245627cf2ff1.png">

The real problem arrises when we try to resize the content - custom buttons are not going into the sidebar.

We are managing to hide them via hacks, or just some css tricks but every time we want to add an additional button, it becomes a struggle and we are having issues with different screen sizes and devices.

We would like to suggest you to build in functionality that allows creation of custom buttons as components.

For Example: 
`<pdf-custom-button
    [icon]="'icon'"
    [label]="'My label'"
    (click)="callmyfunc"
    [styleClass]="'myStyle'"
    [tooltip]="'Text on hover'">
</pdf-custom-button>`

**Comment by stephanrauh** [comment]
That's a really nice idea! Loving it!

Just to make sure we're talking about the same thing: I assume your sidebar is what I call the secondary menu - the kebab menu on the right-hand side?

As for the `<pdf-custom-button>`: Sound's great, but I guess I'll have to wrap my head around the idea. It feels right, but at the moment, I don't know to implement it. Maybe I'll reach out to you again with questions. One thing to consider is the CSS breakpoints. Currently, I've defined three or four screen sizes, and I'd like to enable you to define which button is shifted to the secondary menu on which screen size. I'm not sure if you need this feature, but I guess there are developers out there who need it.

**Comment by salex3** [comment]
Yes!
We are talking about the same sidebar to the right, where all icons hide in it (built-in ones).

Please reach out of you have any questions. The CSS breakpoints actually sound great because developers can define when the buttons show or hide. Suggestions for the breakpoints is to accept either like 'xs', 'md', 'xl' that are going to be the built-in breakpoints OR input 512, or 768, and it'll make it into pixels for the user (that is for very fine tuning).

**Comment by stephanrauh** [comment]
I can't change the breakpoints easily because that'd break the default layout, but I suppose there are enough breakpoints to make almost everybody happy (https://github.com/stephanrauh/ngx-extended-pdf-viewer/blob/3c4dd3e24aa608035acc134825c9907b117fc457/projects/ngx-extended-pdf-viewer/src/lib/dynamic-css/dynamic-css.component.ts#L16):
```css
 public xs = 490;

  public sm = 560;

  public md = 610;

  public lg = 660;

  public xl = 740;

  public xxl = 830;
```
The numbers are pixels. 



Just in case you're looking at the implementation: I've added some voodoo calculating the breakpoints (see https://github.com/stephanrauh/ngx-extended-pdf-viewer/blob/3c4dd3e24aa608035acc134825c9907b117fc457/projects/ngx-extended-pdf-viewer/src/lib/dynamic-css/dynamic-css.component.ts#L176). That's because most embedded PDF files don't have the full width of the window. For example, there might be a side menu. The idea is to make sure the breakpoint triggers when the width of the toolbar exceeds the threshold. In contrast, standard CSS always uses the width of the window to trigger breakpoints, and my workaround fixes this.

**Comment by salex3** [comment]
I was thinking of it to be a custom number which overrides just when the button should be hidden or shown, not to play with the whole breakpoint logic of the viewer.

But if you say that it's not possible, we leave it at that.

**Comment by stephanrauh** [comment]
Counting and measuring buttons sounds like the way to go. It'd allow me to get rid of the CSS magic while offering developers a lot more flexibility. Truth to tell, I was never happy with the breakpoint approach. It works, but it's a far cry from the elegance of the concept of breakpoinst of Bootstrap, Material Design or Tailwind.

If I find a way to provide decent backward compatibility, that's going to be a cool feature for version 16!

**Comment by stephanrauh** [comment]
Your feature has landed with version 18.0.0-alpha.5. I've called the button `<pdf-shy-button>` because it hides in the secondary menu and because every toolbar uses it. Along the way, I've converted several menu entries to shy buttons, so they are now available in the toolbar.

I still need to document the feature. Here's the sneak preview to the documentation. A typical shy button looks like so:

```html
<pdf-shy-button
  [cssClass]="show | responsiveCSSClass : 'always-in-secondary-menu'"
  title="Infinite scroll"
  primaryToolbarId="infiniteScroll"
  l10nId="infinite_scroll"
  [toggled]="pageViewMode == 'infinite-scroll'"
  [action]="onClick"
  l10nLabel="infinite_scroll_label"
  [order]="3400"
  [closeOnClick]="false"
  image="<svg xmlns='http://www.w3.org/2000/svg' height= '24' viewBox= '0 -960 960 960' width='24'><path d= 'M212-260q-90 0-151-65.5T0-482q0-90 61.5-154T212-700q36 0 69.5 12t59.5 37l93 90-42 42-89-87q-18-18-41.5-26t-49.5-8q-64 0-108 46.5T60-482q0 66 43.5 114T212-320q25 0 48.5-8t42.5-25l316-298q26-25 59.5-37t68.5-12q90 0 151.5 64T960-482q0 91-61.5 156.5T747-260q-35 0-69-11.5T619-308l-91-90 42-42 87 87q17 17 41 25t49 8q65 0 109-48t44-114q0-65-44.5-111.5T747-640q-25 0-48.5 9T657-605L341-307q-26 24-60 35.5T212-260Z '/></svg>"
>
</pdf-shy-button>
```

The two `l10n` attributes uses the translation tables of pdf.js. I assume they're useless for you. `[title]` should to the trick just as well.

`[order]` determines where the button appears in the secondary menu. If used increments of 100, so you can insert 99 custom menu items between two default menu items. If the button shows in the toolbar, its position is determined by the position in the source code (plus CSS - in other words, it's the exactly the behavior you'd expect).

The `[action]` attribute is a bit tricky. You can't use `this` in the method because the event listener is used both for the toolbar (`this` is available here) and the secondary menu (`this` is undefined here). I'm sure I've selected a clumsy solution, but here's my current approach:

```ts
export class PdfInfiniteScrollComponent {
  @Output()
  public pageViewModeChange = new EventEmitter<PageViewModeType>();

  public onClick: () => void;

  constructor() {
    const emitter = this.pageViewModeChange;
    this.onClick = () => {
      setTimeout(() => {
           emitter.emit('infinite-scroll');
      });
    };
  }
}
```

I hope I can simplify the code in the next few days. However, the API is already stable.

**Comment by stephanrauh** [comment]
The documentation is now online.
